### PR TITLE
feat: merge develop into modern-cmake

### PR DESCRIPTION
**Pull request by richardapeters**

### CODE REVIEW

File: infra/util/SharedPtr.hpp [hunk 1]
```diff
@@ -16,6 +16,10 @@ namespace infra
     template<class T>
     class WeakPtr;
 
+    extern const struct NoSharedFromThis
+    {
+    } noSharedFromThis;
```
Comment (rjaegers) [inline infra/util/SharedPtr.hpp:21]:
```suggestion
    // clang-format off
    extern const struct NoSharedFromThis {} noSharedFromThis;
    // clang-format on
```

**Review by rjaegers — CHANGES_REQUESTED** [review]

File: hal/interfaces/BackupRam.hpp [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#ifndef HAL_BACKUPRAM_HPP
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply hal/interfaces/BackupRam.hpp:1]: Yes, it is already used in ConnectivityNodes. Let's remove this after the modern-cmake transition.

File: infra/util/SharedPtr.hpp [hunk 3]
```diff
@@ -16,6 +16,8 @@ namespace infra
     template<class T>
     class WeakPtr;
 
+    extern const struct NoSharedFromThis {} noSharedFromThis;
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply infra/util/SharedPtr.hpp]: Done.

File: protobuf/echo/TracingEcho.hpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+#include "services/tracer/Tracer.hpp"
+#include <type_traits>
+
+namespace services
+{
+    class ServiceTracer
+        : public infra::IntrusiveForwardList<ServiceTracer>::NodeType
+    {
+    public:
+        ServiceTracer(uint32_t serviceId);
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply protobuf/echo/TracingEcho.hpp:17]: Done

File: protobuf/echo/TracingEcho.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        EchoOnConnection::SetStreamWriter(tracingWriter.Emplace(std::move(writer), *this));
+    }
+
+    void TracingEchoOnConnection::SendingData(infra::ConstByteRange range) const
+    {
+        infra::ByteInputStream stream(range, infra::noFail);
+        infra::StreamErrorPolicy formatErrorPolicy(infra::softFail);
+        infra::ProtoParser parser(stream, formatErrorPolicy);
+        uint32_t serviceId = static_cast<uint32_t>(parser.GetVarInt());
+        infra::ProtoParser::Field message = parser.GetField();
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply protobuf/echo/TracingEcho.cpp:63]: Done

File: protobuf/protoc_echo_plugin/ProtoCEchoPlugin.hpp [hunk 6]
```diff
@@ -167,6 +167,27 @@ namespace application
         Class* serviceProxyFormatter;
     };
 
+    class TracingServiceGenerator
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply protobuf/protoc_echo_plugin/ProtoCEchoPlugin.hpp:170]: The TracingServiceGenerator is run as part of generating code for TestMessages.proto. That should cover it, right?

File: services/ble/test/CMakeLists.txt [hunk 7]
```diff
@@ -8,9 +8,10 @@ target_link_libraries(services.ble_test PUBLIC
 )
 
 target_sources(services.ble_test PRIVATE
-    TestBondStorageManager.cpp
+    TestBondBlobPersistence.cpp
+    TestBondStorageSynchronizer.cpp
+        TestGap.cpp
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply services/ble/test/CMakeLists.txt:13]: Fixed.

File: services/ble/Gap.hpp [hunk 8]
```diff
@@ -81,6 +82,24 @@ namespace services
         virtual void Standby() = 0;
     };
 
+    class GapPeripheralDecorator
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply services/ble/Gap.hpp:85]: It's a helper class for implementating decorators for GapPeripheral. If you inherit from this, you can add your own decorator code, and use this base class for providing all the default forwarding code

File: services/ble/test/TestGap.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        EXPECT_CALL(gap, SetScanResponseData(infra::ContentsEqual(data)));
+        decorator.SetScanResponseData(data);
+
+        EXPECT_CALL(gap, Advertise(services::GapPeripheral::AdvertisementType::advNonconnInd, 32));
+        decorator.Advertise(services::GapPeripheral::AdvertisementType::advNonconnInd, 32);
+
+        EXPECT_CALL(gap, Standby());
+        decorator.Standby();
+    }
+}
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply services/ble/test/TestGap.cpp:50]: Done

File: protobuf/protoc_echo_plugin/ProtoCEchoPlugin.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (rjaegers) [reply protobuf/protoc_echo_plugin/ProtoCEchoPlugin.hpp:170]: Seeing the Sonar results we should be good.

File: cmake/version.h.in [hunk 10]
```diff
@@ -18,6 +18,7 @@ namespace ${CMAKE_PROJECT_NAME}
     {
         static const char* VERSION EMIL_MAYBE_UNUSED = "${${CMAKE_PROJECT_NAME}_VERSION_STRING}";
         static const char* VERSION_FULL EMIL_MAYBE_UNUSED = "${${CMAKE_PROJECT_NAME}_VERSION_STRING_FULL}";
+        static const char* VERSION_GIT_SHA VARIABLE_MIGHT_BE_UNUSED = "${${CMAKE_PROJECT_NAME}_VERSION_GIT_SHA}";
```
[the remark this answers is not in the document]
Comment (richardapeters) [reply cmake/version.h.in:21]: At its worst, this will result in something like "-NOT_FOUND"

**Review by rjaegers — APPROVED** [review]

**Pull request merged**